code/game/objects/structures/bonfire.dm

```diff
@@ -0,0 +1,180 @@
+		return
+	if(!has_buckled_mobs() && do_after(user, 50, target = src))
+		for(var/obj/item/grown/log/bonfire_log in contents)
+			bonfire_log.forceMove(drop_location())
+			bonfire_log.pixel_x += rand(1,4)
+			bonfire_log.pixel_y += rand(1,4)
+		if(can_buckle || grill)
+			new /obj/item/stack/rods(loc, 1)
+		qdel(src)
+		return
+
+/obj/structure/bonfire/proc/check_oxygen()
+	if(isopenturf(loc))
+		var/turf/open/bonfire_turf = loc
+		if(bonfire_turf.air)
+			var/loc_gases = bonfire_turf.air.gases
+			if(loc_gases[/datum/gas/oxygen] && loc_gases[/datum/gas/oxygen][MOLES] >= 5)
+				return TRUE
+	return FALSE
+
+/obj/structure/bonfire/proc/start_burning()
+	if(burning || !check_oxygen())
+		return
+	icon_state = burn_icon
+	burning = TRUE
+	set_light(6)
+	bonfire_burn()
+	particles = new /particles/bonfire()
+	START_PROCESSING(SSobj, src)
+
+/obj/structure/bonfire/fire_act(exposed_temperature, exposed_volume)
+	start_burning()
+
+/obj/structure/bonfire/proc/on_entered(datum/source, atom/movable/entered)
+	SIGNAL_HANDLER
+	if(burning & !grill)
+		bonfire_burn()
+
+/obj/structure/bonfire/proc/bonfire_burn(delta_time = 2)
+	var/turf/current_location = get_turf(src)
+	if(!grill)
+		current_location.hotspot_expose(1000, 250 * delta_time, 1)
+	for(var/burn_target in current_location)
+		if(burn_target == src)
+			continue
+		else if(isliving(burn_target))
+			var/mob/living/burn_victim = burn_target
+			burn_victim.adjust_fire_stacks(BONFIRE_FIRE_STACK_STRENGTH * 0.5 * delta_time)
+			burn_victim.IgniteMob()
+		else if(isobj(burn_target))
+			var/obj/burned_object = burn_target
+			if(grill && isitem(burned_object))
+				var/obj/item/grilled_item = burned_object
+				SEND_SIGNAL(grilled_item, COMSIG_ITEM_GRILLED, src, delta_time) //Not a big fan, maybe make this use fire_act() in the future.
+				continue
+			burned_object.fire_act(1000, 250 * delta_time)
+
+/obj/structure/bonfire/process(delta_time)
+	if(!check_oxygen())
+		extinguish()
+		return
+	bonfire_burn(delta_time)
+
+/obj/structure/bonfire/extinguish()
+	if(burning)
+		icon_state = "bonfire"
+		burning = FALSE
+		set_light(0)
+		QDEL_NULL(particles)
+		STOP_PROCESSING(SSobj, src)
+
+/obj/structure/bonfire/buckle_mob(mob/living/buckled_mob, force = FALSE, check_loc = TRUE)
+	if(..())
+		buckled_mob.pixel_y += 13
+
+/obj/structure/bonfire/unbuckle_mob(mob/living/buckled_mob, force=FALSE)
+	if(..())
+		buckled_mob.pixel_y -= 13
+
+/particles/bonfire
+	icon = 'icons/effects/particles/bonfire.dmi'
+	icon_state = "bonfire"
+	width = 100
+	height = 100
+	count = 1000
+	spawning = 4
+	lifespan = 0.7 SECONDS
+	fade = 1 SECONDS
+	grow = -0.01
+	velocity = list(0, 0)
+	position = generator("circle", 0, 16, NORMAL_RAND)
+	drift = generator("vector", list(0, -0.2), list(0, 0.2))
+	gravity = list(0, 0.95)
+	scale = generator("vector", list(0.3, 0.3), list(1,1), NORMAL_RAND)
+	rotation = 30
+	spin = generator("num", -20, 20)
```

Comment:
also, look at all the code above. an in game particle tester for coders would be monumental